Refactoring usage of `forall_symbols` macro into c++11 loops and replacement of asserts.

thomasspriggs opened the pull request:
This refactors usage of the `forall_symbols` macro into c++11 loops instead. Some variable renaming to match coding standards also included.

thomasspriggs committed:
Refactor `forall_symbols` usage into c++11 loop.

thomasspriggs committed:
Remove unused macro `forall_symbol_module_map`

thomasspriggs committed:
Remove unused `forall_symbols` macro.

thomasspriggs committed:
Replace asserts with new equivalents.

tautschnig commented:
Can "Variable renaming for coding standards." please be explained to me?

Edit: I suppose that's "Avoid abbreviations (e.g. prefer `symbol_table` to `st`)." -- but why is "temp" less of an abbreviation than "tmp"?

If doing variable renamings is on the agenda, then I'd really suggest to put this in a different pull request.

hannes-steffenhagen-diffblue commented on src/goto-instrument/alignment_checks.cpp:
```diff
@@ -19,21 +19,21 @@ void print_struct_alignment_problems(
-          it_mem!=components.end();
+      for(struct_typet::componentst::const_iterator it_mem = components.begin();
+          it_mem != components.end();
           it_mem++)
```
Why all these whitespace changes? Makes it harder to see what are actual code changes. If cleaning up the formatting is a good idea, can you do this in a separate commit?

thomasspriggs replied:
Because I ran my updates through clang-format before making each commit. I now realise these should have been kept in separate commits.

owen-mc-diffblue replied:
@hannes-steffenhagen-diffblue If you add "?w=1" to the end of the URL when viewing a diff on github then it ignores whitespace when making the diff.

tautschnig replied:
Add to that a later use of `git blame` also requires an additional `-w`. So, yes, there are workarounds, but maybe the pain can be avoided in the first place?

owen-mc-diffblue replied:
I agree that running clang-format should be done in a separate commit in this instance, as it will make the PR much easier to review. Ignoring whitespace when viewing a diff is still useful in other situations, e.g. when a conditional is introduced which changes the indentation of a large section of code.

thk123 replied:
Though with that flag you can no longer add comments seemingly

thomasspriggs commented:
> Can "Variable renaming for coding standards." please be explained to me?

The coding standards state we should "Avoid abbreviations" in naming. The variable renaming is expanding abbreviations into non abbreviated names. Examples include -
- "dest" -> "destination"
- "c" -> "code"
- "op" -> "operands"
edit: I shall move variable renaming into a separate PR.

tautschnig commented:
So cleanup is good. But it's really helpful if different kinds of cleanup can be put in different pull requests, and if whitespace cleanup can be factored into different commits. Being conscious of reviewing is vital for making reviews viable and useful.

thomasspriggs committed:
Fix formatting in assert replacements.

martin-cs commented:
If we are going to do this then we should also take out the other custom forall_... macros, which looks to me like a wide, shallow change that doesn't actually improve things other than a subjective notion of "clarity" or "cleanliness".

LAJW commented:
@martin-cs It brings the codebase closer to the "vanilla" C++ - which prefers templates and lambdas over macros. Makes the codebase much more readable to new developers, just because they don't have to look up every possible variant of `for_each_<your type goes here>` macros. I was convinced for the longest time that we were doing something unique with those macros (like non-standard, recursive iteration), but then realised they were just less smart version of `BOOST_FOREACH` (which is no longer useful in C++11).

martin-cs commented:
@LAJW Which issue does this fix?  Which new feature depends on it?  All of these "clean-ups" and "change style to ..." have a cost; what I'm not seeing is the benefit.

LAJW commented:
@martin-cs If you require quantifiable benefits apart from improving readability and reducing potential future errors (which is the primary concern of this PR), Ranged-for loops enable us to use const references rather than mutable pointers to mutable data (iterators). Const has proven to allow compilers to optimise the program even further.

As to address the problem of "inconsistency" of having many different styles in many different places - I think it's important to improve the code gradually, especially since we've seen time and time again backlash against massive PRs replacing every instance of something in the codebase ending up with those PRs not getting merged or taking months to do so. I much prefer inconsistent code with some good parts over consistently awful code (which is what we will continue having to deal with if we actively discourage PRs like this).

hannes-steffenhagen-diffblue commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2489,6 +2489,7 @@ codet java_bytecode_convert_methodt::convert_instructions(
           INVARIANT(
             a_it2->second.stack.size() == stack.size(),
             "Stack sizes should be the same.");
+          stackt::const_iterator os_it=a_it2->second.stack.begin();
```
This shouldn't be in the formatting commit

thk123 replied:
This introduces a variable the last commit deleted and should be squashed into the appropriate commit.

thomasspriggs replied:
I have now moved this fix into the correct commit.

thk123 commented on src/goto-instrument/alignment_checks.cpp:
```diff
@@ -19,21 +19,21 @@ void print_struct_alignment_problems(
   std::ostream &out)
 {
-  forall_symbols(it, symbol_table.symbols)
-    if(it->second.is_type && it->second.type.id()==ID_struct)
```
Good riddance to this confusing indentation!

thk123 commented on src/goto-programs/class_hierarchy.cpp:
```diff
@@ -55,28 +54,29 @@ void class_hierarchyt::operator()(const symbol_tablet &symbol_table)
         const irep_idt &parent = to_symbol_type(base.type()).get_identifier();
         if(!parent.empty())
-          add_edge(nodes_by_name.at(parent), nodes_by_name.at(it->first));
+          add_edge(
```
Maybe addressed in a later commit, but since `add_edge` is now spread over multiple lines, please wrap in braces

thk123 commented on src/goto-programs/goto_convert_functions.cpp:
```diff
@@ -38,16 +38,14 @@ void goto_convert_functionst::goto_convert(goto_functionst &functions)
+      symbol_pair.second.type.id() == ID_code &&
+      (symbol_pair.second.mode == ID_C || symbol_pair.second.mode == ID_cpp ||
+       symbol_pair.second.mode == ID_java || symbol_pair.second.mode == "jsil"))
+      symbol_list.push_back(symbol_pair.first);
```
Was fairly unreadable before, but possibly even more so now, can we wrap `symbol_list.push_back` in braces (if not done by a later commit already)

thk123 commented on src/goto-programs/goto_convert_functions.cpp:
```diff
@@ -59,12 +57,12 @@ void goto_convert_functionst::goto_convert(goto_functionst &functions)
+    if(!symbol_pair.second.is_type &&
+       symbol_pair.second.type.id()==ID_code &&
+       symbol_pair.second.value.is_not_nil())
+      symbol_pair.second.value=codet();
```
Ditto

thk123 commented on src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2486,8 +2486,9 @@ codet java_bytecode_convert_methodt::convert_instructions(
         else
         {
-          assert(a_it2->second.stack.size()==stack.size());
-          stackt::const_iterator os_it=a_it2->second.stack.begin();
```
This variable is disappearing, should it be? It probably shouldn't in this commit in any case

thomasspriggs commented:
FYI - I am intending to fix the issues with this PR, but I have other work the takes higher priority for the moment.

thomasspriggs committed:
Add missing brackets to multiline if statements.

thomasspriggs commented:
I think I have addressed the comments which were asking for code updates. Please re-review. I have not addressed the combination of white space updates with code changes in the same commit, because I don't have a way to separate these out without re-work. I will bear this in mind for future and keep reformatting in separate commits.

thomasspriggs commented:
@thk123 Please re-review.

tautschnig commented:
> I have not addressed the combination of white space updates with code changes in the same commit, because I don't have a way to separate these out without re-work.

My apologies in case I'm stating the obvious, or indeed my approach of doing this kind of work may be far from optimal:
```
git rebase -i
# mark the white-space-mix commit for editing
# keep a backup of the commit message if desired
git reset --soft HEAD^
git reset HEAD .
git add -p
# selectively add desired changes, and possibly split & edit some of the hunks to achieve the separation
git commit
git diff
# review what's left and possibly commit
git rebase --continue
```

tautschnig reviewed (approved):
I'm ok with this. Is it necessary? I don't know, and I can certainly side with @martin-cs' points. Somehow we need to find a good middle ground of not blocking modernisation while also not putting unbounded amounts of effort on teams that have patch stacks outside the code base.
